pdf/pdfselection.cpp

```diff
@@ -0,0 +1,433 @@
+        *boxIndex = 0;
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+QPointF PDFSelection::start() const
+{
+    if (!d->canvas || d->pageIndexStart < 0)
+        return QPointF();
+    
+    return d->canvas->fromPageToItem(d->pageIndexStart, d->start);
+}
+
+void PDFSelection::setStart(const QPointF &point)
+{
+    int pageIndex, boxIndex;
+    d->textBoxId(point, PDFSelection::Private::After, &pageIndex, &boxIndex);
+    if (pageIndex < 0 || boxIndex < 0
+        || (pageIndex == d->pageIndexStart && boxIndex == d->boxIndexStart)) {
+        return;
+    }
+    if (pageIndex > d->pageIndexStop
+        || (pageIndex == d->pageIndexStop && boxIndex > d->boxIndexStop)) {
+        return;
+    }
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(pageIndex);
+    QRectF box = boxes[boxIndex].first;
+
+    int nBoxes = d->sliceCount(pageIndex, boxIndex, d->pageIndexStart, d->boxIndexStart);
+    if (nBoxes > 0) {
+        beginInsertRows(QModelIndex(), 0, nBoxes - 1);
+        d->pageIndexStart = pageIndex;
+        d->boxIndexStart = boxIndex;
+        endInsertRows();
+    } else {
+        beginRemoveRows(QModelIndex(), 0, -nBoxes - 1);
+        d->pageIndexStart = pageIndex;
+        d->boxIndexStart = boxIndex;
+        endRemoveRows();
+    }
+
+    d->start = QPointF(box.x(), box.y() + box.height() / 2.);
+    emit startChanged();
+
+    emit countChanged();
+    emit textChanged();
+}
+
+QPointF PDFSelection::stop() const
+{
+    if (!d->canvas || d->pageIndexStop < 0)
+        return QPointF();
+    
+    return d->canvas->fromPageToItem(d->pageIndexStop, d->stop);
+}
+
+void PDFSelection::setStop(const QPointF &point)
+{
+    int pageIndex, boxIndex;
+    d->textBoxId(point, PDFSelection::Private::Before, &pageIndex, &boxIndex);
+    if (pageIndex < 0 || boxIndex < 0
+        || (pageIndex == d->pageIndexStop && boxIndex == d->boxIndexStop)) {
+        return;
+    }
+    if (pageIndex < d->pageIndexStart
+        || (pageIndex == d->pageIndexStart && boxIndex < d->boxIndexStart)) {
+        return;
+    }
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(pageIndex);
+    QRectF box = boxes[boxIndex].first;
+
+    int count_ = count(); 
+    int nBoxes = d->sliceCount(d->pageIndexStop, d->boxIndexStop, pageIndex, boxIndex);
+    if (nBoxes > 0) {
+        beginInsertRows(QModelIndex(), count_, count_ + nBoxes - 1);
+        d->pageIndexStop = pageIndex;
+        d->boxIndexStop = boxIndex;
+        endInsertRows();
+    } else {
+        beginRemoveRows(QModelIndex(), count_ + nBoxes, count_ - 1);
+        d->pageIndexStop = pageIndex;
+        d->boxIndexStop = boxIndex;
+        endRemoveRows();
+    }
+
+    d->stop = QPointF(box.x() + box.width(), box.y() + box.height() / 2.);
+    emit stopChanged();
+
+    emit countChanged();
+    emit textChanged();
+}
+
+void PDFSelection::selectAt(const QPointF &point)
+{
+    unselect();
```

Comment:
Fine as is, but optimally there wouldn't be unnecessary resets or inserts. Now with long press this does first reset even if there wasn't a selection. Protecting unselect() for that should be trivial. Rarer case of long press while selection exists could only do one reset to update state, but that complicates the code and not necessarily worth it.